src/datachain/query/dataset.py

```diff
@@ -1220,6 +1223,28 @@ def _order_columns(
     return [[d[n] for n in column_order] for d in column_dicts]
 
 
+def _remap_sys_id_dense(
+    columns: Iterable[ColumnElement], branch_offset: int
+) -> list[ColumnElement]:
+    """Return a new column list where sys__id is remapped to a dense row number.
+
+    The resulting sys__id is computed as:
+      row_number(order by sys__id) * 2 + branch_offset
+
+    This guarantees uniqueness across branches (even vs odd) and avoids
+    arithmetic overflow from scaling arbitrary pre-existing ids.
+    """
+    remapped: list[ColumnElement] = []
+    for col in columns:
+        if getattr(col, "name", None) == "sys__id":
+            rn = sqlalchemy.func.row_number().over(order_by=col)
+            new_id = (rn * 2) + sqlalchemy.literal(branch_offset)
```

Comment:
Just wonder, should be have `2` as function argument with default value?

Something like:
```python
def _remap_sys_id_dense(
    columns: Iterable[ColumnElement],
    branch_offset: int,
    branches_total: int = 2,
) -> list[ColumnElement]:
```

to be able to use more than 2 branches in future.

Otherwise we might want to use `bool` as `branch_offset` (like `even`/`odd` branch), since only two branches are supported here.